Allow to configure Quartz misfire handling strategy

rmanibus opened the pull request:
fix #24254

rmanibus committed:
Allow to configure Quartz misfire handling strategy

rmanibus commented:
cc @machi1990

machi1990 commented on extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzRuntimeConfig.java:
```diff
@@ -45,4 +49,22 @@
     public QuartzStartMode startMode;
+
+    /**
+     * per job configuration.
```
```suggestion
     * Misfire policy per job configuration.
```

machi1990 commented on extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzRuntimeConfig.java:
```diff
@@ -45,4 +49,22 @@
+     * per job configuration.
+     */
+    @ConfigDocSection
+    @ConfigDocMapKey("job-identity")
```
```suggestion
    @ConfigDocMapKey("identity")
```

machi1990 commented on extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzRuntimeConfig.java:
```diff
@@ -45,4 +49,22 @@
+    @ConfigDocSection
+    @ConfigDocMapKey("job-identity")
+    @ConfigItem(name = ConfigItem.PARENT)
+    public Map<String, QuartzNamedRuntimeConfig> namedJobs;
```
```suggestion
    public Map<String, QuartzNamedRuntimeConfig> misfirePolicyPerJobs;
```

machi1990 commented on extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java:
```diff
@@ -174,15 +181,66 @@ public QuartzScheduler(SchedulerContext context, QuartzSupport quartzSupport, Sc
                             }
-                            scheduleBuilder = CronScheduleBuilder.cronSchedule(cron);
+                            CronScheduleBuilder cronScheduleBuilder = CronScheduleBuilder.cronSchedule(cron);
+                            if (runtimeConfig.namedJobs.containsKey(identity)) {
```
```suggestion
                            if (runtimeConfig.misfirePolicyPerJobs.containsKey(identity)) {
```

machi1990 replied:
Maybe we could just do a `.get(identity)` and check for null. 
This way we'd have the value of the policy already which we could reuse throughout.
What do you think?

machi1990 commented on extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java:
```diff
@@ -174,15 +181,66 @@ public QuartzScheduler(SchedulerContext context, QuartzSupport quartzSupport, Sc
-                            scheduleBuilder = CronScheduleBuilder.cronSchedule(cron);
+                            CronScheduleBuilder cronScheduleBuilder = CronScheduleBuilder.cronSchedule(cron);
+                            if (runtimeConfig.namedJobs.containsKey(identity)) {
+                                switch (runtimeConfig.namedJobs.get(identity).misfirePolicy) {
```
```suggestion
                                switch (runtimeConfig.misfirePolicyPerJobs.get(identity).misfirePolicy) {
```

machi1990 commented on extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java:
```diff
@@ -174,15 +181,66 @@ public QuartzScheduler(SchedulerContext context, QuartzSupport quartzSupport, Sc
+                                        SIMPLE_TRIGGER_RESCHEDULE_NOW_WITH_REMAINING_REPEAT_COUNT,
+                                        SIMPLE_TRIGGER_RESCHEDULE_NEXT_WITH_EXISTING_COUNT,
+                                        SIMPLE_TRIGGER_RESCHEDULE_NEXT_WITH_REMAINING_COUNT)
+                                        .contains(runtimeConfig.namedJobs.get(identity).misfirePolicy)) {
```
```suggestion
                                        .contains(runtimeConfig.misfirePolicyPerJobs.get(identity).misfirePolicy)) {
```

machi1990 commented on extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java:
```diff
@@ -174,15 +181,66 @@ public QuartzScheduler(SchedulerContext context, QuartzSupport quartzSupport, Sc
+                            SimpleScheduleBuilder simpleScheduleBuilder = SimpleScheduleBuilder.simpleSchedule()
                                     .withIntervalInMilliseconds(everyMillis.getAsLong())
                                     .repeatForever();
+                            if (runtimeConfig.namedJobs.containsKey(identity)) {
```
```suggestion
                            if (runtimeConfig.misfirePolicyPerJobs.containsKey(identity)) {
```

machi1990 commented on extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java:
```diff
@@ -174,15 +181,66 @@ public QuartzScheduler(SchedulerContext context, QuartzSupport quartzSupport, Sc
                                     .withIntervalInMilliseconds(everyMillis.getAsLong())
                                     .repeatForever();
+                            if (runtimeConfig.namedJobs.containsKey(identity)) {
+                                switch (runtimeConfig.namedJobs.get(identity).misfirePolicy) {
```
```suggestion
                                switch (runtimeConfig.misfirePolicyPerJobs.get(identity).misfirePolicy) {
```

machi1990 commented on extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java:
```diff
@@ -174,15 +181,66 @@ public QuartzScheduler(SchedulerContext context, QuartzSupport quartzSupport, Sc
+                                        simpleScheduleBuilder.withMisfireHandlingInstructionNextWithRemainingCount();
+                                        break;
+                                }
+                                if (CRON_TRIGGER_DO_NOTHING.equals(runtimeConfig.namedJobs.get(identity).misfirePolicy)) {
```
```suggestion
                                if (CRON_TRIGGER_DO_NOTHING.equals(runtimeConfig.misfirePolicyPerJobs.get(identity).misfirePolicy)) {
```

machi1990 reviewed (changes requested):
This is great work.
Thank you.
I've left some comments, very minor and mostly around renaming.
Would you also mind a section in the docs? Thank you.

machi1990 commented on extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/MisfirePolicyTest.java:
```diff
@@ -0,0 +1,161 @@
+            .withApplicationRoot((jar) -> jar
+                    .addClasses(Jobs.class)
+                    .addAsResource(new StringAsset(
+                            "quarkus.quartz.\"simple_ignore_misfire_policy\".misfire-policy=IGNORE_MISFIRE_POLICY\n" +
```
```suggestion
                            "quarkus.quartz.\"simple_ignore_misfire_policy\".misfire-policy=ignore-misfire-policy\n" +
```
The enums are hyphenated, so you could actually do as suggested. 
This can be applied throughout.

Looking at the test, I propose we change the order of the configuration key, to properly set the context. So instead of reading like `quarkus.quartz."<identity>.misfire-policy"`, it will read like `quarkus.quartz.misfire-policy."<identity>"`. I find the later to be easier to understand that we are dealing with the misfire configuration, especially when you have long job identity. 

The configuration could be something like
```java
/**
     * Misfire policy per job configuration.
     */
    @ConfigDocSection
    @ConfigDocMapKey("identity")
    @ConfigItem(name = "misfire-policy")
    public Map<String, QuartzNamedRuntimeConfig> misfirePolicyPerJob;

    @ConfigGroup
    public static class QuartzNamedRuntimeConfig {
        /**
         * The quartz misfire policy for this job.
         */
        @ConfigItem(defaultValue = "SMART_POLICY", name = ConfigItem.PARENT)
        public QuartzMisfirePolicy misfirePolicy;
    }
```

rmanibus replied:
Sounds good, but what if we have to add more per job parameters in the future ?

machi1990 replied:
The parameters will be prefixed by what we are looking to expose. Just like what I propose here.

machi1990 commented on extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/MisfirePolicyTest.java:
```diff
@@ -0,0 +1,161 @@
+import io.quarkus.scheduler.Scheduler;
+import io.quarkus.test.QuarkusUnitTest;
+
+public class MisfirePolicyTest {
```
Thanks for adding the tests.

machi1990 commented on extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java:
```diff
@@ -174,15 +181,66 @@ public QuartzScheduler(SchedulerContext context, QuartzSupport quartzSupport, Sc
+                                        SIMPLE_TRIGGER_RESCHEDULE_NEXT_WITH_EXISTING_COUNT,
+                                        SIMPLE_TRIGGER_RESCHEDULE_NEXT_WITH_REMAINING_COUNT)
+                                        .contains(runtimeConfig.namedJobs.get(identity).misfirePolicy)) {
+                                    LOGGER.warnf(
+                                            "configured cron job %s with invalid misfire policy. defaulting to SMART_POLICY",
+                                            identity);
+                                }
```
I am wondering if throwing an exception is actually better.

machi1990 replied:
we could also explain/show the possible configuration for this type of job

rmanibus commented:
> This is great work.
> Thank you.
> I've left some comments, very minor and mostly around renaming.
> Would you also mind a section in the docs? Thank you.

@machi1990 I made the changes and added some doc.

machi1990 commented on extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzRuntimeConfig.java:
```diff
@@ -45,4 +49,22 @@
+        /**
+         * The quartz misfire policy for this job.
+         */
+        @ConfigItem(defaultValue = "SMART_POLICY", name = ConfigItem.PARENT)
```
```suggestion
        @ConfigItem(defaultValue = "smart-policy", name = ConfigItem.PARENT)
```

machi1990 commented on docs/src/main/asciidoc/quartz.adoc:
```diff
@@ -133,10 +133,20 @@ public class TaskBean {
 }
 ----
+
+[source,conf]
+----
+quarkus.quartz.misfire-policy.task-job=ignore-misfire-policy <5>
+----
```
Maybe this can go in this part https://github.com/quarkusio/quarkus/blob/3f1926ae23d4ea1276669b0f8dd960d73feee6ee/docs/src/main/asciidoc/quartz.adoc#updating-the-application-configuration-file

machi1990 commented on docs/src/main/asciidoc/quartz.adoc:
```diff
@@ -133,10 +133,20 @@ public class TaskBean {
 <4> Persist the task in database using xref:hibernate-orm-panache.adoc[Panache].
+<5> The misfire policy can be configured for each job.
+
+Valid misfire policy for cron jobs are: smart-policy, ignore-misfire-policy, fire-now and cron-trigger-do-nothing.
```
```suggestion
Valid misfire policy for cron jobs are: `smart-policy`, `ignore-misfire-policy`, `fire-now` and `cron-trigger-do-nothing`.
```

machi1990 commented on docs/src/main/asciidoc/quartz.adoc:
```diff
@@ -133,10 +133,20 @@ public class TaskBean {
+<5> The misfire policy can be configured for each job.
+
+Valid misfire policy for cron jobs are: smart-policy, ignore-misfire-policy, fire-now and cron-trigger-do-nothing.
+Valid misfire policy for interval jobs are: smart-policy, ignore-misfire-policy, fire-now, simple-trigger-reschedule-now-with-existing-repeat-count, simple-trigger-reschedule-now-with-remaining-repeat-count, simple-trigger-reschedule-next-with-existing-count and simple-trigger-reschedule-next-with-remaining-count.
```
```suggestion
Valid misfire policy for interval jobs are: `smart-policy`, `ignore-misfire-policy`, `fire-now`, `simple-trigger-reschedule-now-with-existing-repeat-count`, `simple-trigger-reschedule-now-with-remaining-repeat-count`, `simple-trigger-reschedule-next-with-existing-count` and `simple-trigger-reschedule-next-with-remaining-count`.
```

machi1990 reviewed (approved):
lgtm

Thanks for the awesome work @rmanibus

mkouba commented on docs/src/main/asciidoc/quartz.adoc:
```diff
@@ -201,6 +201,7 @@ Edit the `application.properties` file and add the below configuration:
 # Quartz configuration
 quarkus.quartz.clustered=true <1>
 quarkus.quartz.store-type=jdbc-cmt <2>
+quarkus.quartz.misfire-policy.task-job=ignore-misfire-policy <3>
```
The `task-job` is a job identity, or?

machi1990 replied:
Yes it is the job identity. 
@rmanibus whenever you get the time, would you mind mentioning that in the doc as well.

gsmet reviewed (changes requested):
Please don't merge already, I will make minor adjustments to the exception messages tomorrow.

gsmet committed:
Quartz misfire policy - Better error messages

Two changes here:
- always handle of the cases in the enum so that we get proper warnings
  if a value is not taken into account
- use the actual values that we document as valid values in the error
  messages

gsmet reviewed (approved):
@rmanibus I added an additional commit for you to look at.

gsmet commented:
Thanks!
